openshift-knative-operator/pkg/common/deployment.go

```diff
@@ -0,0 +1,73 @@
+package common
+
+import (
+	mf "github.com/manifestival/manifestival"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// InjectEnvironmentIntoDeployment injects the specified environment variables into the
+// specified deployment/container.
+// Note: We're not deleting empty environment variables and instead set them to empty
+// string. Three-way-merging of the deployment drops the update otherwise.
+func InjectEnvironmentIntoDeployment(deploymentName, containerName string, env map[string]string) mf.Transformer {
+	return transformDeployment(deploymentName, func(deploy *appsv1.Deployment) error {
+		containers := deploy.Spec.Template.Spec.Containers
+		for i := range containers {
+			c := &containers[i]
+			if c.Name != containerName {
+				continue
+			}
+
+			for key, value := range env {
+				c.Env = upsert(c.Env, key, value)
+			}
+		}
+
+		return nil
+	})
+}
+
+// upserts updates the env var if the key already exists or inserts it if it didn't
+// exist.
+func upsert(orgEnv []corev1.EnvVar, key, value string) []corev1.EnvVar {
+	// Set the value if the key is already present.
+	for i := range orgEnv {
+		if orgEnv[i].Name == key {
+			orgEnv[i].Value = value
+			return orgEnv
+		}
+	}
+	// If not, append a key-value pair.
+	return append(orgEnv, corev1.EnvVar{
+		Name:  key,
+		Value: value,
+	})
+}
+
+// transformDeployment returns a transformer that transforms a deployment with the given
+// name.
+func transformDeployment(name string, f func(*appsv1.Deployment) error) mf.Transformer {
+	return func(u *unstructured.Unstructured) error {
+		if u.GetKind() != "Deployment" || u.GetName() != name {
```

Comment:
Should not this be an AND condition?

Reply:
Note that this exits early. So we want to exit early if the Kind is not a deployment OR if it is and the name does not match.

Reply:
I understand the short-cut condition but we might accidentally modify the wrong deployment if the kind is right but the name is not. Basically if it is a deployment we ignore the name no?

Reply:
No we don't. If it's a Deployment the first condition will be false, so the second one will be evaluated which will be true (because the name doesn't match), so it's true overall.

Reply:
Right. So it is equivelant to ` u.GetKind() == "Deployment" && u.GetName() == name` afaik.

Reply:
Right. Since we are transforming a Deployment (name of the method) we should return quickly if the kind is not a deployment anyway, not check the name.

Reply:
Right.